### DISCUSSION TITLE
PyInstaller, PySide, MacOS and Accessability

### DISCUSSION THREAD

**Discussion by ajax-dunkin-v** [pyinstaller/pyinstaller#8705] [PyInstaller] [answered]
Hello, everyone.

I have app on PySide2 and deploy in with PyInstaller 3.6.
I want to update PyInstaller to the latest version, the deployment goes without errors. But after deploying with the latest version of PyInstaller on MacOS, the accessibility IDs stopped showing (via accessibility inspector)

MacOS versions I tried: Ventura, Sonoma and MacOS 15 (beta)
Python==3.9.13 (installed from python.org)
PySide2==5.15.2.1

The test application is very primitive

main.py
```
import sys
import os

from PySide2.QtGui import QGuiApplication
from PySide2.QtQml import QQmlApplicationEngine


if getattr(sys, 'frozen', False):
    bundle_dir = sys._MEIPASS
else:
    bundle_dir = os.path.dirname(os.path.abspath(__file__))


app = QGuiApplication(sys.argv)

engine = QQmlApplicationEngine()
engine.quit.connect(app.quit)
engine.load(os.path.join(bundle_dir, "main.qml"))

sys.exit(app.exec_())

```

main.qml
```
import QtQuick 2.12
import QtQuick.Controls 2.12


ApplicationWindow {
    visible: true 
    width: 600
    height: 500
    title: "MyApp"

    Rectangle {
        anchors.fill: parent 

        Text {
            id: my_text

            anchors.centerIn: parent 
            text: "My APP"
            font.pixelSize: 45

            Accessible.role: Accessible.StaticText
            Accessible.name: my_text.text
            Accessible.description: "my app text"
        }

        Accessible.role: Accessible.StaticText
        Accessible.name: "Rectagle"
        Accessible.description: "my app rectangle"
    }
}

```

If I deploy this example with PyInstaller 3.6, the Accessibility Inspector will show that accessibility ids exist for the text box

<img width="929" alt="Screenshot 2024-08-02 at 13 57 37" src="https://github.com/user-attachments/assets/2a44f65c-df4c-4840-b54d-3a629bd2f523">

If I deploy this example with PyInstaller 6.9, the Accessibility Inspector will show that accessibility id only for main window

<img width="943" alt="Screenshot 2024-08-02 at 14 04 24" src="https://github.com/user-attachments/assets/4494de81-4d3b-4234-b4a5-0d06541f1028">

Command for deploy
`pyinstaller --clean main.py -wn 'MyApp' --exclude-module tkinter --noconfirm --add-data main.qml:. --osx-bundle-identifier MyTestIdent`

For PyInstaller 3.6 don't forget remove signature for Python
`codesign --remove-signature --deep dist/MyApp.app/Contents/MacOS/Python`

Can anyone help me with this issue?

**Comment by rokm** [comment]
Do you perform some other action with the .app bundle after building it? E.g., copy or move it, or compress, transfer, uncompress it? I.e., something that might break symlinks in it? (Because .app bundles made with PyInstaller 6.x depend heavily on symlinks...).

I can't reproduce this. If I build your example with python 3.9.13, PyInstaller 6.9.0, and PySide2 5.15.2.1 on x86_64 Monterey system, the resulting .app bundle seems to set the accessibility values as expected, both on the same Moneterey system, and on arm64 Sonoma system (the screenshots are taken on the latter):

![img1](https://github.com/user-attachments/assets/2c893b57-6cda-42db-b130-809d8a9387ee)

![img2](https://github.com/user-attachments/assets/0f5832f1-c23c-479a-96c0-8bda05c5b0a0)

**Comment by ajax-dunkin-v** [comment]
Thanks for the answer.
No, I do not perform any other actions with the bundle.
Yes, when building on Monterey, everything works. And on Sonoma and Ventura.
To reproduce the problem, it is necessary to deploy it on Sonoma

**Reply by rokm** [reply]
>  To reproduce the problem, it is necessary to deploy it on Sonoma

You mean *to build* it on Sonoma?

**Reply by rokm** [reply]
And just to confirm - this is using x86_64 environment (even if OS is arm64), because PySide2 does not have arm64 PyPI wheels? (Or are you building your own?)

**Reply by rokm** [reply]
Also, which python 3.9.13 variant - macos10.9 (the "intel-only installer") or macos11 (the "universal2 installer")?

**Reply by ajax-dunkin-v** [reply]
Yes i mean _build_.

Enviroment - x86_64 (under rosetta).
Python - "intell-only" installer

**Reply by rokm** [reply]
OK, I'll give it a try later today.

**Reply by rokm** [reply, marked as answer]
Hmmm, yeah, build made with intel-only (macos10.9) variant of python 3.9.13 and PyInstaller 6.9.0 indeed does not work as expected under Sonoma.

But to be fair, neither does running your example as unfrozen script in that python environment.

This boils down to the macOS SDK declared by the executable; in that macos10.9 (intel-only) python 3.9.13 build, python executable and shared library declare SDK version 10.9.

Starting with PyInstaller 4.4 (#5839), we change the reported SDK version on the generated frozen executable to match the one from the python executable (if lower), in order to ensure consistent behavior between the unfrozen python and frozen application (and so, you are getting that *consistent* behavior now).

You can see this in the build log, e.g.,
```
9305 INFO: Rewriting the executable's macOS SDK version (13.1.0) to match the SDK version of the Python library (10.9.0) in order to avoid inconsistent behavior and potential UI issues in the frozen application.
```

The example I made earlier on Monterey system was, in fact, using the macos11 (universal2 installer) python, so in that case, the SDK version was down-graded only to 12.1.

The pre-built bootloader in PyInstaller 3.6 was build with SDK 10.11, and since that PyInstaller version does not modify SDK version on the frozen executable, that is the reported version.

It seems that the minimum required SDK version for these accessibility descriptions to work correctly is 10.10.

You have three options:

1. you can switch from intel-only python installer to universal2 one, and run the x86_64 slice via Rosetta.

2. you can disable [this part](https://github.com/pyinstaller/pyinstaller/blob/d068e06887caea5ded22b48789dac7e76379bd86/PyInstaller/building/api.py#L853-L871) of code in your local copy of PyInstaller

3. you can re-adjust the SDK version on the created executable in post-processing, e.g.,

```python
from PyInstaller.utils import osx as osxutils
print(osxutils.get_macos_sdk_version('dist/MyApp.app/Contents/MacOS/MyApp'))  # Check that version is indeed 10.9.0
osxutils.set_macos_sdk_version('dist/MyApp.app/Contents/MacOS/MyApp', 10, 11, 0)
```

(You'll probably want to re-sign the .app bundle after modification, even if just with ad-hoc signature).

**Reply by ajax-dunkin-v** [reply]
Thank you!
Yes, I also tried to rebuild with universal python - everything worked.
Thanks for the options, I will choose an acceptable solution.
Have a nice day)
